proposals/transcoding_invalid_identifiers/README.md

```diff
@@ -0,0 +1,335 @@
+
+## Differences to Punycode
+
+A summary of the differences against Punycode are:
+- The separating character changes from `-` to `_`. Since `-` is invalid character in OpenUSD.
+- Delta encoding starts from `0` instead of `128`. This is to account the fact not all ASCII's are allowed in OpenUSD.
+- The base representation changes from `36` in Punycode to `62` in this implementation (to represent more information
+in less characters).
+- Threshold is constant in this implementation. There is no loss of performance or memory representation, since our 
+`base` is also increased, and simplifies implementation.
+
+# Examples
+
+The above example:
+
+| Original        | Transcoding           |
+|-----------------|-----------------------|
+| 012-345-678/9.0 | tn__01234567890_lG7QQ |
+
+```cpp
+static_assert(
+    SdfBoostringEncodeIdentifier("012-345-678/9.0") == "tn__01234567890_lG7QQ",
+);
+static_assert(
+    SdfBootstringDecodeIdentifier("tn__01234567890_lG7QQ") == "012-345-678/9.0",
+);
+```
+
+No changes.
+
+| Original      | Transcoding   |
+|---------------|---------------|
+| id12345_abcde | id12345_abcde |
+
+```cpp
+static_assert(
+    SdfBoostringEncodeIdentifier("id12345_abcde") == "id12345_abcde",
+);
+static_assert(
+    SdfBootstringDecodeIdentifier("id12345_abcde") == "id12345_abcde",
+);
+```
+
+Remove invalid characters. The extended characters is `-` and `/`.
+
+| Original    | Transcoding       |
+|-------------|-------------------|
+| 123-456/555 | tn__123456555_oDT |
+
+```cpp
+static_assert(
+    SdfBoostringEncodeIdentifier("123-456/555") == "tn__123456555_oDT",
+);
+static_assert(
+    SdfBootstringDecodeIdentifier("tn__123456555_oDT") == "123-456/555",
+);
+```
+
+Convert UTF-8 characters to valid ASCII (i.e. `TfMakeValidIdentifier`). This can be useful to share identifiers
+between new versions of OpenUSD and legacy versions. The extended character set is `ü`, `,` and ` `(space).
+
+| Original         | Transcoding               |
+|------------------|---------------------------|
+| München, Germany | tn__MnchenGermany_pDV5hi2 |
+
+
+```cpp
+static_assert(
+    SdfBoostringEncodeAsciiIdentifier("München, Germany") == "tn__MnchenGermany_pDV5hi2",
+);
+static_assert(
+    SdfBootstringDecodeIdentifier("tn__MnchenGermany_pDV5hi2") == "München, Germany",
+);
+```
+
+Convert UTF-8 characters to valid XID. The extended character set is, `,` and ` `(space). Notice how the decoding 
+function is the same as above (i.e. `SdfBootstringDecodeIdentifier`).
+
+| Original         | Transcoding             |
+|------------------|-------------------------|
+| München, Germany | tn__MünchenGermany_rEi5 |
+
+
+```cpp
+static_assert(
+    SdfBoostringEncodeIdentifier("München, Germany") == "tn__MünchenGermany_rEi5",
+);
+static_assert(
+    SdfBootstringDecodeIdentifier("tn__MünchenGermany_rEi5") == "München, Germany",
+);
+```
+
+Send invalid strings to encoding/decoding functions should result in the original string.
```

Comment:
What if it `std::optional` was used instead?  Let the user decide how to handle.

```
SdfBootstringDecodeIdentifier("tn__///abc") == std::optional<std::string>{};
```
If you wanted to replicate that behavior you could always do this-- but you wouldn't have to.
```
std::string decoded = SdfBoostringDecodeIdentifier("tn__mycoolstring").value_or("tn__mycoolstring");
```

Reply:
Oh yeap, I like that. I keep forgetting about optional in C++.

Reply:
Added the optional in the API and examples. 

I also broke the previous examples of `coolstring`. There is a difference between encoding those (they are actually valid,. so I talk about them in `Encoding valid identifiers produces no changes.`) vs decoding them.